### PR TITLE
New vector invariant scheme

### PR DESCRIPTION
**Pull request by simone-silvestri**
This PR introduces full upwinding in vector invariant advection, which allows running eddy-permitting/resolving hydrostatic simulations without the need for an explicit viscous dissipation.

In addition, it also refactors the whole vector scheme superseding PR #2820 

Examples of surface speed in modified neverworld configuration:

https://user-images.githubusercontent.com/33547697/206742914-d216ea2b-7cd2-4aa5-be70-a760b088698b.mp4

left:      full upwinding 1/8th of a degree, no horizontal closure
center: full upwinding 1/4th of a degree, no horizontal closure
right:    conservative vector invariant, biharmonic viscosity

Closes #2820

### CODE REVIEW

File: src/Advection/vector_invariant_advection.jl [hunk 1]
```diff
@@ -1,86 +1,180 @@
...
+const VectorInvariantConserving = Union{VectorInvariantEnergyConserving, VectorInvariantEnstrophyConserving}
 
-@inline U_dot_∇u(i, j, k, grid, scheme::VectorInvariantSchemes, U) = (
-    + vertical_vorticity_U(i, j, k, grid, scheme, U.u, U.v)  # Vertical relative vorticity term
-    + vertical_advection_U(i, j, k, grid, scheme, U.u, U.w)  # Horizontal vorticity / vertical advection term
-    + bernoulli_head_U(i, j, k, grid, scheme, U.u, U.v))     # Bernoulli head term
+@inline U_dot_∇u(i, j, k, grid, scheme::VectorInvariant, U) = (
+    + horizontal_advection_U(i, j, k, grid, scheme, U.u, U.v)      
+    + vertical_advection_U(i, j, k, grid, scheme, U.w, U.u)
+    + bernoulli_head_U(i, j, k, grid, scheme, U.u, U.v))
```
Comment (navidcy) [inline src/Advection/vector_invariant_advection.jl:61]:
```suggestion
    + horizontal_advection_U(i, j, k, grid, scheme, U.u, U.v)
    + vertical_advection_U(i, j, k, grid, scheme, U.w, U.u)
    + bernoulli_head_U(i, j, k, grid, scheme, U.u, U.v))
```

File: src/Advection/vector_invariant_advection.jl [hunk 2]
```diff
@@ -1,86 +1,180 @@
...
+    + horizontal_advection_U(i, j, k, grid, scheme, U.u, U.v)      
+    + vertical_advection_U(i, j, k, grid, scheme, U.w, U.u)
+    + bernoulli_head_U(i, j, k, grid, scheme, U.u, U.v))              
     
-@inline U_dot_∇v(i, j, k, grid, scheme::VectorInvariantSchemes, U) = (
-    + vertical_vorticity_V(i, j, k, grid, scheme, U.u, U.v)  # Vertical relative vorticity term
-    + vertical_advection_V(i, j, k, grid, scheme, U.v, U.w)  # Horizontal vorticity / vertical advection term
-    + bernoulli_head_V(i, j, k, grid, scheme, U.u, U.v))     # Bernoulli head term
+@inline U_dot_∇v(i, j, k, grid, scheme::VectorInvariant, U) = (
+    + horizontal_advection_V(i, j, k, grid, scheme, U.u, U.v)
```
Comment (navidcy) [inline src/Advection/vector_invariant_advection.jl:64]:
```suggestion
    + horizontal_advection_V(i, j, k, grid, scheme, U.u, U.v)
```

File: src/Advection/vector_invariant_advection.jl [hunk 3]
```diff
@@ -96,22 +190,26 @@ end
...
+        @inline inner_right_biased_interpolate_xᶠᵃᵃ(i, j, k, grid, scheme::UX{$buffer}, f::Function, idx, loc, VI::AbstractSmoothnessStencil, args...) = inner_right_biased_interpolate_xᶠᵃᵃ(i, j, k, grid, scheme, f, idx, loc, args...)
+        @inline inner_right_biased_interpolate_yᵃᶠᵃ(i, j, k, grid, scheme::U{$buffer},  f::Function, idx, loc, VI::AbstractSmoothnessStencil, args...) = inner_right_biased_interpolate_yᵃᶠᵃ(i, j, k, grid, scheme, f, idx, loc, args...)
+        @inline inner_right_biased_interpolate_yᵃᶠᵃ(i, j, k, grid, scheme::UY{$buffer}, f::Function, idx, loc, VI::AbstractSmoothnessStencil, args...) = inner_right_biased_interpolate_yᵃᶠᵃ(i, j, k, grid, scheme, f, idx, loc, args...)
+        @inline inner_right_biased_interpolate_zᵃᵃᶠ(i, j, k, grid, scheme::U{$buffer},  f::Function, idx, loc, VI::AbstractSmoothnessStencil, args...) = inner_right_biased_interpolate_zᵃᵃᶠ(i, j, k, grid, scheme, f, idx, loc, args...)
+        @inline inner_right_biased_interpolate_zᵃᵃᶠ(i, j, k, grid, scheme::UZ{$buffer}, f::Function, idx, loc, VI::AbstractSmoothnessStencil, args...) = inner_right_biased_interpolate_zᵃᵃᶠ(i, j, k, grid, scheme, f, idx, loc, args...)
+    end
+end
```
Comment (navidcy) [inline src/Advection/vector_invariant_advection.jl:215]:
```suggestion
end

```

File: src/Advection/weno_reconstruction.jl [hunk 4]
```diff
@@ -30,16 +25,16 @@ struct WENO{N, FT, XT, YT, ZT, VI, WF, PP, CA, SI} <: AbstractUpwindBiasedAdvect
...
-    function WENO{N, FT, VI, WF}(coeff_xᶠᵃᵃ::XT, coeff_xᶜᵃᵃ::XT,
-                                 coeff_yᵃᶠᵃ::YT, coeff_yᵃᶜᵃ::YT, 
-                                 coeff_zᵃᵃᶠ::ZT, coeff_zᵃᵃᶜ::ZT,
-                                 bounds::PP, buffer_scheme::CA,
-                                 advecting_velocity_scheme :: SI) where {N, FT, XT, YT, ZT, VI, WF, PP, CA, SI}
+    function WENO{N, FT, WF}(coeff_xᶠᵃᵃ::XT, coeff_xᶜᵃᵃ::XT,
+                             coeff_yᵃᶠᵃ::YT, coeff_yᵃᶜᵃ::YT, 
+                             coeff_zᵃᵃᶠ::ZT, coeff_zᵃᵃᶜ::ZT,
+                             bounds::PP, buffer_scheme::CA,
+                             advecting_velocity_scheme :: SI) where {N, FT, XT, YT, ZT, VI, WF, PP, CA, SI}
```
Comment (navidcy) [inline src/Advection/weno_reconstruction.jl:32]:
```suggestion
                             advecting_velocity_scheme :: SI) where {N, FT, XT, YT, ZT, WF, PP, CA, SI}
```

File: src/Advection/vector_invariant_advection.jl [hunk 5]
```diff
@@ -1,86 +1,180 @@
...
+    
+    function VectorInvariant{N, FT}(vorticity_scheme::Z, divergence_scheme::D, vorticity_stencil::ZS, divergence_stencil::DS, vertical_scheme::V) where {N, FT, Z, D, ZS, DS, V}
+        return new{N, FT, Z, D, ZS, DS, V}(vorticity_scheme, divergence_scheme, vorticity_stencil, divergence_stencil, vertical_scheme)
+    end
+end
 
-struct VectorInvariant{S, CA}
-    scheme :: S
-    buffer_scheme :: CA
+function VectorInvariant(; vorticity_scheme::AbstractAdvectionScheme{N, FT} = EnstrophyConservingScheme(),
```
Comment (navidcy) [inline src/Advection/vector_invariant_advection.jl:21]: add a docstring? just so we can figure out what are the available options here?

**Review by glwagner — APPROVED** [review]

File: Project.toml [hunk 6]
```diff
@@ -1,7 +1,7 @@
 name = "Oceananigans"
 uuid = "9e8cae18-63c1-5223-a75c-80ca9d6e9a09"
 authors = ["Climate Modeling Alliance and contributors"]
-version = "0.78.6"
+version = "0.78.7"
```
Comment (navidcy) [inline Project.toml:4]: wasn't it breaking change?

**Review by navidcy — APPROVED** [review]

File: Project.toml (hunk 6 above)
Comment (simone-silvestri) [reply Project.toml:4]: Ah ok, I misunderstood

**Comment by navidcy** [comment]
merge?

**Pull request merged**